### PR TITLE
Enhance image size validation logic

### PR DESCRIPTION
**Pull request by mthalman**
Updates Image Builder's logic for image size validation so that it can detect when extraneous baseline data exists and support updating baseline data only for images whose size is out of the allowed range.  

These changes basically required a complete rewrite of the logic.  The new changes have split the logic into two separate commands, `validateImageSize` and `updateImageSizeBaseline`, because they represent two distinct operations and have separate options.  They do have some shared logic so that's been factored into a common base class `ImageSizeCommandBase`.

For testability, the `ValidateImageSizeCommand` class exposes a data structure that represents the output it displays in the console.

Related to https://github.com/dotnet/dotnet-docker/issues/936

### CODE REVIEW

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/ImageSizeCommandBase.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        where TOptions : ImageSizeOptionsBase, new()
+    {
+        public ImageSizeCommandBase(IDockerService dockerService)
+        {
+            this.DockerService = dockerService ?? throw new ArgumentNullException(nameof(dockerService));
+        }
+
+        protected IDockerService DockerService { get; }
+
+        protected void ProcessImages(ImageHandler processImage)
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/ImageSizeCommandBase.cs]: It provides context as to what values need to be passed in by naming the parameters.

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/UpdateImageSizeBaselineOptions.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{
+    public class UpdateImageSizeBaselineOptions : ImageSizeOptionsBase
+    {
+        protected override string CommandHelp => "Updates an image size baseline file with current image sizes";
+
+        public bool OutOfRangeOnly { get; set; }
+
+        public override void ParseCommandLine(ArgumentSyntax syntax)
+        {
+            bool outOfRangeOnly = false;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/UpdateImageSizeBaselineOptions.cs]: In that case, I propose changing the name.  I prefer options having a default value of false.  So how about an option named `all`, with a default value of false, to indicate that all baselines should be updated?

File: src/Microsoft.DotNet.ImageBuilder/src/IEnvironmentService.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.DotNet.ImageBuilder
+{
+    public interface IEnvironmentService
+    {
+        void Exit(int exitCode);
+    }
+}
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/IEnvironmentService.cs]: This always bites me when using the "Extract Interface" refactoring option.  I wish VS would be smart enough to do this automatically.

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/UpdateImageSizeBaselineOptions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/UpdateImageSizeBaselineOptions.cs]: I agree the default value should be false.  I like you name suggestion as well.

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/ValidateImageSizeOptions.cs [hunk 4]
```diff
@@ -6,41 +6,23 @@
...
         public override void ParseCommandLine(ArgumentSyntax syntax)
         {
-            base.ParseCommandLine(syntax);
-
-            FilterOptions.ParseCommandLine(syntax);
-
-            int allowedVariance = 5;
-            syntax.DefineOption("variance", ref allowedVariance, $"Allowed percent variance in size (default is `{allowedVariance}`");
-            AllowedVariance = allowedVariance;
+            bool checkNewOrOldImagesOnly = false;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/ValidateImageSizeOptions.cs]: I'll be honest that I didn't like the name either. It's definitely not easy to come up with something short yet descriptive.  I like your suggestion but how about something like `BaselineIntegrityOnly` to indicate that it's not an additional check but a more restrictive check?

[the remark this answers is not in the document]
Comment (MichaelSimons) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/ValidateImageSizeOptions.cs]: I like your suggestion better.

File: src/Microsoft.DotNet.ImageBuilder/src/Commands/ImageSizeCommandBase.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.ImageBuilder.ViewModel;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.DotNet.ImageBuilder.Commands
+{
+    public delegate void ImageHandler(string repoId, string imageId, string imageTag);
+
+    public abstract class ImageSizeCommandBase<TOptions> : ManifestCommand<TOptions>
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Commands/ImageSizeCommandBase.cs]: No, it's not necessary.  I've removed the suffix.

**Pull request merged**